tests/asgi/tests.py

```diff
@@ -402,3 +426,95 @@ class TestASGIHandler(ASGIHandler):
+        self.assertEqual(response_body["type"], "http.response.body")
+        self.assertEqual(response_body["body"], b"Hello World!")
+        await communicator.wait()
+        self.assertIs(view_did_cancel, False)
+
+        # Request cycle with a disconnect.
+        application = TestASGIHandler()
+        scope = self.async_request_factory._base_scope(path="/cancel/")
+        communicator = ApplicationCommunicator(application, scope)
+        await communicator.send_input({"type": "http.request"})
+        response_start = await communicator.receive_output()
+        # Fetch the start of response so streaming can begin
+        self.assertEqual(response_start["type"], "http.response.start")
+        self.assertEqual(response_start["status"], 200)
+        await asyncio.sleep(0.1)
+        # Now disconnect the client.
+        await communicator.send_input({"type": "http.disconnect"})
+        # This time the handler should not send a response.
+        with self.assertRaises(asyncio.TimeoutError):
+            await communicator.receive_output()
+        await communicator.wait()
+        self.assertIs(view_did_cancel, True)
+
+    async def test_streaming(self):
+        factory = AsyncRequestFactory()
+        scope = factory._base_scope(path="/streaming/", query_string=b"sleep=0.001")
+        application = get_asgi_application()
+        communicator = ApplicationCommunicator(application, scope)
+        await communicator.send_input({"type": "http.request"})
+
+        # Fetch headers.
+        await communicator.receive_output(timeout=1)
+        # Fetch the 'foo' and 'bar'.
+        first_response = await communicator.receive_output(timeout=1)
+        second_response = await communicator.receive_output(timeout=1)
+        self.assertEqual(first_response["body"], b"foo\n")
+        self.assertEqual(second_response["body"], b"bar\n")
+
+        # Fetch the rest of the response so that coroutines are cleaned up.
+        await communicator.receive_output(timeout=1)
+        with self.assertRaises(asyncio.TimeoutError):
+            await communicator.receive_output(timeout=1)
+
+    async def test_streaming_disconnect(self):
+        factory = AsyncRequestFactory()
+        scope = factory._base_scope(path="/streaming/", query_string=b"sleep=0.1")
+        application = get_asgi_application()
+        communicator = ApplicationCommunicator(application, scope)
+        await communicator.send_input({"type": "http.request"})
+        await communicator.receive_output(timeout=1)
+        first_response = await communicator.receive_output(timeout=1)
+        self.assertEqual(first_response["body"], b"foo\n")
+        # Simulate disconnection.
+        await communicator.send_input({"type": "http.disconnect"})
+        # Ensure that 'bar\n' isn't sent.
+        with self.assertRaises(asyncio.TimeoutError):
+            await communicator.receive_output(timeout=0.2)  # type: ignore
```

Comment:
I can see `ResourceWarning` when running `asgi` tests on Python 3.12:
```
$ python -Wall ./runtests.py asgi.tests.ASGITest --parallel=1
Testing against Django installed in '/django/django'
Found 20 test(s).
System check identified no issues (0 silenced).
....................
----------------------------------------------------------------------
Ran 20 tests in 8.087s

OK
Exception ignored in: <function SpooledTemporaryFile.__del__ at 0x7f66ad4b8540>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/tempfile.py", line 736, in __del__
    _warnings.warn(
ResourceWarning: Unclosed file <tempfile.SpooledTemporaryFile object at 0x7f66ab688cd0>
```

Reply:
There is no typing in Django, so this suppression comment can be removed.

```suggestion
            await communicator.receive_output(timeout=0.2)
```